cxx/gendb.cc

```diff
@@ -0,0 +1,157 @@
+             bool _only_final_emissions, bool _record_class_is_clean)
+    : schema(schema_),
+      schema_helper(schema_, _only_final_emissions, _record_class_is_clean) {
+  std::map<std::string, std::vector<std::string>>
+      annotated_domains_for_relation;
+  T_schema hirm_schema =
+      schema_helper.make_hirm_schema(&annotated_domains_for_relation);
+  hirm = new HIRM(hirm_schema, prng);
+
+  for (const auto& [class_name, unused_class] : schema.classes) {
+    domain_crps[class_name] = CRP();
+  }
+}
+
+void GenDB::incorporate(
+    std::mt19937* prng,
+    const std::pair<int, std::map<std::string, ObservationVariant>>& row) {
+  int id = row.first;
+
+  // Maps a query relation name to an observed value.
+  std::map<std::string, ObservationVariant> vals = row.second;
+
+  // Loop over the values in the row, and their query relation names.
+  for (const auto& [query_rel, val] : vals) {
+    // Sample a set of items to be incorporated into the query relation.
+    const std::vector<std::string>& class_path =
+        schema.query.fields.at(query_rel).class_path;
+    std::vector<int> items =
+        sample_entities_relation(prng, schema.query.record_class,
+                                 class_path.cbegin(), class_path.cend(), id);
+
+    // Incorporate the items/value into the query relation.
+    incorporate_query_relation(prng, query_rel, items, val);
+  }
+}
+
+// This function walks the class_path of the query, populates the global
+// reference_values table if necessary, and returns a sampled set of items
+// for the query relation.
+std::vector<int> GenDB::sample_entities_relation(
+    std::mt19937* prng, const std::string& class_name,
+    std::vector<std::string>::const_iterator class_path_start,
+    std::vector<std::string>::const_iterator class_path_end, int class_item) {
+  if (class_path_end - class_path_start == 1) {
+    // These are domains and we need to DFS-traverse the class's
+    // parents, similar to PCleanSchemaHelper::compute_domains_for.
+    return sample_class_ancestors(prng, class_name, class_item);
+  } else {
+    // These are noisy relation domains along the path from the latent cleanly-
+    // observed class to the record class.
+    std::string ref_field = *class_path_start;
+
+    // If the reference field isn't populated, sample a value from a CRP and
+    // add it to reference_values.
+    std::string ref_class =
+        std::get<ClassVar>(
+            schema.classes.at(class_name).vars.at(ref_field).spec)
+            .class_name;
+    if (!reference_values[class_name].contains(class_item)) {
+      sample_and_incorporate_reference(prng, class_name, class_item, ref_field,
+                                       ref_class);
+    }
+    std::vector<int> items = sample_entities_relation(
+        prng, ref_class, ++class_path_start, class_path_end,
+        reference_values[class_name][class_item][ref_field]);
+    items.push_back(class_item);
```

Comment:
Maybe worth a comment here about how when a relation is generated from class C, we always put the domain C last in the relation's list of domains, which is why the class_item goes last in the list of items.

Reply:
Done.